src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -48,60 +49,99 @@ internal static SafeChannelBindingHandle QueryChannelBinding(SafeSslHandle conte
-                // close-notify, you'll probably need to rewrite SafeSslHandle.Disconnect().
-                // https://www.openssl.org/docs/manmaster/ssl/SSL_shutdown.html
-                Ssl.SslCtxSetQuietShutdown(innerContext);
+                IntPtr method = GetSslMethod(protocols);
 
-                if (!Ssl.SetEncryptionPolicy(innerContext, policy))
+                using (SafeSslContextHandle innerContext = Ssl.SslCtxCreate(method))
                 {
-                    throw new PlatformNotSupportedException(SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
-                }
+                    if (innerContext.IsInvalid)
+                    {
+                        throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    }
 
-                if (certHandle != null && certKeyHandle != null)
-                {
-                    SetSslCertificate(innerContext, certHandle, certKeyHandle);
-                }
+                    // Configure allowed protocols. It's ok to use DangerousGetHandle here without AddRef/Release as we just
+                    // create the handle, it's rooted by the using, no one else has a reference to it, etc.
+                    Ssl.SetProtocolOptions(innerContext.DangerousGetHandle(), protocols);
 
-                if (remoteCertRequired)
-                {
-                    Debug.Assert(isServer, "isServer flag should be true");
-                    Ssl.SslCtxSetVerify(innerContext,
-                        s_verifyClientCertificate);
+                    // The logic in SafeSslHandle.Disconnect is simple because we are doing a quiet
+                    // shutdown (we aren't negotiating for session close to enable later session
+                    // restoration).
+                    //
+                    // If you find yourself wanting to remove this line to enable bidirectional
+                    // close-notify, you'll probably need to rewrite SafeSslHandle.Disconnect().
+                    // https://www.openssl.org/docs/manmaster/ssl/SSL_shutdown.html
+                    Ssl.SslCtxSetQuietShutdown(innerContext);
 
-                    //update the client CA list 
-                    UpdateCAListFromRootStore(innerContext);
-                }
+                    if (!Ssl.SetEncryptionPolicy(innerContext, policy))
+                    {
+                        throw new PlatformNotSupportedException(SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
+                    }
 
-                context = SafeSslHandle.Create(innerContext, isServer);
-                Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
-                if (context.IsInvalid)
-                {
-                    context.Dispose();
-                    throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    if (hasCertificateAndKey)
```

Comment:
I'll look at the change in more depth later, but just for now, note that DangerousAddRef throws an exception if it can't add the ref (which is fine, that exception can just propagate) which means checks like this shouldn't be necessary.